File: cpp/prob004.cpp
#include <iostream>
#include <chrono>
#include "euler.h"

#define PROB_NUM    4

int compute()
{
    uint32_t num = 0;
    
    for(int i = 999; i >= 100; i--){
        for(int j = 999; j >= 100; j--){
            uint32_t x = i*j;
            if(is_palindrome(x) && num < x) num = x;
        }
    }
    return num;
}


int test_answer(std::string prob_answer) {
    std::string answer;
    if(!get_answer(PROB_NUM, &answer)) {
        std::cout << "file not found" << std::endl;
        return 1;
    }

    assert(answer.compare(prob_answer) == 0);
    return 0;
}

int main(int argc, char const *argv[])
{
    auto t1 = std::chrono::high_resolution_clock::now();
    int64_t result = compute();
    auto t2 = std::chrono::high_resolution_clock::now();

    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1).count();
    std::cout << result << "\tin " << duration << " us\n";

    // do testing
    test_answer(std::to_string(result));
    return 0;
}

File: rust/src/bin/prob004.rs
extern crate euler;

fn compute() -> i32{
    let mut max_palin = 0;
    for i in (99..1000).rev() {
        for j in (99..1000).rev(){
            let p = i*j;
            if p < max_palin { break }
            
            if euler::is_palindrome(p) {
                max_palin = p;
                break;
            }
        }
    }

    max_palin
}

fn main() {
    let now = std::time::Instant::now();
    let result = compute();
    println!("{:?}\tin {}ms", result, now.elapsed().as_millis());
}


#[cfg(test)]
mod tests {
    use super::*;
    const PROB_NUM: usize = 4;
    
    #[test]
    fn test_prob004() {
        let result = compute().to_string();
        euler::test_answer(PROB_NUM, result);
    }
}